### PR TITLE
Added talkback support for button accessibility: disabled prop

### PR DESCRIPTION
**Pull request by huzaifaaak**
## Summary

Issue # https://github.com/facebook/react-native/issues/30934 .When using a screen reader disabled buttons do not announce that they are disabled.

## Changelog

[Android] [Changed] - Passing accessibility state in button so it can announce disabled in talkback

## Test Plan

I have added Button in Button Example with accessibiltyState prop that will announce button is disabled when testing with talkback.

## Ios test
I am unable to run ios project on my machine. RNTesterPods.xcworkspace gives workspace integrity error :/

### CODE REVIEW

**Comment by huzaifaaak** [comment]
@kacieb Can you please go review this and give your comments 🙏

File: Libraries/Components/Button.js [hunk 1]
```diff
@@ -273,11 +278,14 @@ class Button extends React.Component<ButtonProps> {
         buttonStyles.push({backgroundColor: color});
       }
     }
-    const accessibilityState = {};
+    const accessibilityState =
+      disabled != null
```
Comment (lunaleaps) [inline Libraries/Components/Button.js:282]: Alternatively we could do `const accessibilityState = disabled === true ? ... : ...` so it's `accessiblityState.disabled | disabled`.

File: packages/rn-tester/js/examples/Modal/ModalExample.js [hunk 2]
```diff
@@ -172,10 +172,14 @@ class ModalExample extends React.Component<{...}, $FlowFixMeState> {
             this.setState({currentOrientation: evt.nativeEvent.orientation})
           }
           onDismiss={() => {
-            if (this.state.action === 'onDismiss') alert(this.state.action);
+            if (this.state.action === 'onDismiss') {
```
Comment (lunaleaps) [inline packages/rn-tester/js/examples/Modal/ModalExample.js:175]: Can we move this to a separate PR for formatting?

File: packages/rn-tester/js/examples/Button/ButtonExample.js [hunk 3]
```diff
@@ -174,6 +174,29 @@ exports.examples = [
       );
     },
   },
+  {
+    title: 'Button with accessibilityState="disabled"',
```
Comment (lunaleaps) [inline packages/rn-tester/js/examples/Button/ButtonExample.js:178]: Nit: accessibilityState={{disabled:true}}

File: Libraries/Components/Button.js (hunk 1 above)
Comment (lunaleaps) [inline Libraries/Components/Button.js:282]: I'd suggest `const accessibilityState = disabled === true ? ... : ...` so it's `accessiblityState.disabled | disabled`.

File: packages/rn-tester/js/examples/Modal/ModalExample.js (hunk 2 above)
Comment (lunaleaps) [inline packages/rn-tester/js/examples/Modal/ModalExample.js:175]: Can we move all the formatting to a separate PR?

**Comment by lunaleaps** [comment]
Thanks for adding an example as well!

File: packages/rn-tester/js/examples/Modal/ModalExample.js (hunk 2 above)
Comment (huzaifaaak) [reply packages/rn-tester/js/examples/Modal/ModalExample.js:175]: @lunaleaps I think the linting command added the formatting. I haven't changed anything.

Comment (huzaifaaak) [reply packages/rn-tester/js/examples/Modal/ModalExample.js:175]: Should I manually revert it to the old one ?

Comment (huzaifaaak) [reply packages/rn-tester/js/examples/Modal/ModalExample.js]: @lunaleaps formatting reverted 😄

File: Libraries/Components/Button.js [hunk 4]
```diff
@@ -18,7 +18,7 @@ const Text = require('../Text/Text');
 const TouchableNativeFeedback = require('./Touchable/TouchableNativeFeedback');
 const TouchableOpacity = require('./Touchable/TouchableOpacity');
 const View = require('./View/View');
-
+import type {AccessibilityState} from './View/ViewAccessibility';
```
Comment (lunaleaps) [inline Libraries/Components/Button.js:21]: Nit: should this also be grouped with line 24? I can also fix this on import

**Review by lunaleaps — APPROVED** [review]
Looks good to me!

File: Libraries/Components/Button.js (hunk 4 above)
Comment (huzaifaaak) [reply Libraries/Components/Button.js:21]: Yes, this should be grouped with other typed imports 💯. Please fix this during import

File: Libraries/Components/Button.js [hunk 5]
```diff
@@ -273,11 +278,14 @@ class Button extends React.Component<ButtonProps> {
...
-    const accessibilityState = {};
+    const accessibilityState =
+      disabled === true
+        ? {...this.props.accessibilityState, disabled}
+        : this.props.accessibilityState;
+
     if (disabled) {
       buttonStyles.push(styles.buttonDisabled);
       textStyles.push(styles.textDisabled);
-      accessibilityState.disabled = true;
```
Comment (lunaleaps) [inline Libraries/Components/Button.js:280]: @kacieb brought up a great point -- this PR doesn't seem to be fixing accessibility for talkback but rather adding an ability for accessibilityState -- I tested on an Android device and did notice that talkback does announce "disabled" for the button in your example where you solely use `accessibilityState`. 

That raises the question of what was the original issue reporting? Was it reporting that `accessibilityState={{disabled..` wasn't working -- which is indeed what your PR fixes -- but the fact that the property didn't even exist before has me puzzled

Comment (lunaleaps) [reply Libraries/Components/Button.js:280]: cc @blavalla who reported the issue for clarification -- relevant to our discussion on disabled vs. accessibilityState.disabled

Comment (lunaleaps) [reply Libraries/Components/Button.js:280]: I'm going to hold off landing this PR until we have some clarity about what the original issue is.

Comment (huzaifaaak) [reply Libraries/Components/Button.js:280]: Noted! 
Link to the accessibility issue: https://github.com/facebook/react-native/issues/30840

Comment (blavalla) [reply Libraries/Components/Button.js:280]: @lunaleaps, the original issue was that setting `accessibilityState: disabled` alone was not actually disabling the button for accessibility.

This snack (https://snack.expo.io/d1Ikg7nqo) from #30840 shows the issue well.

Comment (blavalla) [reply Libraries/Components/Button.js:280]: To clarify further, the issue was mostly that it is weird that you can actually set conflicting states for accessibility and non-accessibility users.  Basically, there is nothing stopping someone from setting `disabled="true"` `accessibilityState={{disabled: false}}` or vice-versa. 

If we are going to explicitly allow this though, they should at least work as expected, with the accessibilityState: disabled actually disabling the element for accessibility users.

Comment (lunaleaps) [reply Libraries/Components/Button.js:280]: Ah right! `accessibilityState` is a View level prop so it should already exist on `<Button />`. My mistake! Okay that clarifies things! It sounds like this change is still worth pursuing -- to just sync the two values

Comment (huzaifaaak) [reply Libraries/Components/Button.js:280]: > If we are going to explicitly allow this though, they should at least work as expected, with the accessibilityState: disabled actually disabling the element for accessibility users.

@lunaleaps I explicitly tested `disabled="false"` and  `accessibilityState={{disabled: true}}` and it **announces** that button is **disabled** but it **does not disables** the **onPress**. I believe this needs to be done on the **View** level.

Having two different props is confusing. To disable the onPress we have to `explicitly` pass the `disabled={true}` prop.

Comment (lunaleaps) [reply Libraries/Components/Button.js:280]: > and it announces that button is disabled but it does not disables the onPress. I believe this needs to be done on the View level.

Yea I think that's covered by https://github.com/facebook/react-native/issues/30943

Yea we can at least move forward with this to add both and we are discussing about consolidating the two props

File: Libraries/Components/Button.js [hunk 6]
```diff
@@ -273,11 +279,14 @@ class Button extends React.Component<ButtonProps> {
         buttonStyles.push({backgroundColor: color});
       }
     }
-    const accessibilityState = {};
+
+    if (disabled != null) {
+      accessibilityState.disabled = disabled;
+    } //Should works for both disabled={true/false}
```
Comment (huzaifaaak) [inline Libraries/Components/Button.js:285]: typo: work

**Comment by huzaifaaak** [comment]
@lunaleaps @kacieb I have updated this pr's code. It's now giving `disabled prop` the **authority**. It automatically set's up `accessibilityState={disabled}`. Also, I added up #30943 by disabling the onpress through `accessibilityState.disabled ` if components own disabled prop is missing.
Can you please give your comments?

**Comment by lunaleaps** [comment]
Will look over this later today, thanks @huzaifaaak for the update!

File: Libraries/Components/Button.js [hunk 7]
```diff
@@ -299,7 +308,7 @@ class Button extends React.Component<ButtonProps> {
         nextFocusRight={nextFocusRight}
         nextFocusUp={nextFocusUp}
         testID={testID}
-        disabled={disabled}
+        disabled={disabled != null ? disabled : accessibilityState?.disabled} //This will sync accessibilityState={disabled} with disabled prop.
```
Comment (lunaleaps) [inline Libraries/Components/Button.js:311]: I think this should be done on line 286? Otherwise we're missing the logic in that if statement if accessibilityState.disabled is true

Comment (huzaifaaak) [reply Libraries/Components/Button.js:311]: I was a bit confused on this. For now it's not appyling the `disable style` if only accessibilityState?.disabled is passed.

Comment (huzaifaaak) [reply Libraries/Components/Button.js:311]: I think this needs to be passed here and on line 286 so it applies style and disables the button aswell if accessibilityState.disabled is true

Comment (lunaleaps) [reply Libraries/Components/Button.js:311]: Yea I think we should treat the button as disabled if accessibilityState.disabled is true in all manners --- behaviour and styling

Comment (kacieb) [reply Libraries/Components/Button.js:311]: > Yea I think we should treat the button as disabled if accessibilityState.disabled is true in all manners --- behaviour and styling

I think I agree with you -I think if accessibilityState.disabled=true and disabled=false though, disabled should take precedence. If disabled is undefined, then I agree accessibilityState.disabled=true should disable the component.

Comment (huzaifaaak) [reply Libraries/Components/Button.js]: @lunaleaps @kacieb thank you for the explanation. I have updated the accessibilityState.disabled=true to disable styling aswell.

File: Libraries/Components/Button.js (hunk 1 above)
Comment (lunaleaps) [reply Libraries/Components/Button.js:287]: Sorry to follow up on this can we avoid creating a new object here?
```
const accessibilityState = disabled !== accessibilityState.disabled ? {...this.props.accessibilityState, disabled} : this.props.accessibilityState
```

**Review by lunaleaps — APPROVED** [review]
Otherwise looks good to me

**Comment by huzaifaaak** [comment]
@lunaleaps I have updated it to avoid object creation. Do let me know if it can be done in a better way

**Comment by kacieb** [comment]
This is awesome, thank you so much for working on this and improving React Native's accessibility! I'd love to see a snapshot test for this component at some point, similar to this Pressable one added [here](https://github.com/facebook/react-native/commit/1c7d9c8046099eab8db4a460bedc0b2c07ed06df), that ensures the `accessibilityState` values are correctly set as expected.

**Comment by amarlette** [comment]
Hey @huzaifaaak I would like to give you a shout-out on Twitter and include you in our end-of-month issues update for your contribution. Do you have a Twitter we can tag and is your @ the same as your Github?

**Comment by huzaifaaak** [comment]
Hey @amarlette yes my twitter handle is the same [huzaifaaak](https://twitter.com/huzaifaaak)
